File: 例程/兵王问题SVM程序/testSVMChess/testSVMChessLibSVM.m
% fid  =  fopen('krkopt.data'); % Read data
% c = fread(fid, 3); % Useless data
% 
% vec = zeros(6,1); % Store position
% xapp = []; % Data sets
% yapp = []; % Labels
% while ~feof(fid)
%     string = [];
%     c = fread(fid, 1); % Read 1 byte
%     while c ~= 13 % c != '\x0d'
%         string = [string, c];
%         c = fread(fid,1); % Read 1 byte
%     end
%     fread(fid, 1); % Read '\x0a'
%     if length(string) > 10
%         vec(1) = string(1) - 96; % Letter 1
%         vec(2) = string(3) - 48; % Number 1
%         vec(3) = string(5) - 96; % Letter 2
%         vec(4) = string(7) - 48; % Number 2
%         vec(5) = string(9) - 96; % Letter 3
%         vec(6) = string(11) - 48; % Number 3
%         xapp = [xapp, vec];
%         if string(13) == 100 % 'draw'
%             yapp = [yapp, 1];
%         else
%             yapp = [yapp, -1];
%         end
%     end
% end
% fclose(fid);
% 
% [N, M] = size(xapp);
% p = randperm(M); % 直接打乱了训练样本
% numberOfSamplesForTraining = 5000; % Number of training samples
% xTraining = [];
% yTraining = [];
% for i = 1 : numberOfSamplesForTraining
%     xTraining = [xTraining, xapp(:, p(i))]; % Get training data sets
%     yTraining = [yTraining, yapp(p(i))]; % Get training labels
% end
% xTraining = xTraining'; % Transpose
% yTraining = yTraining'; % Transpose
% 
% xTesting = [];
% yTesting = [];
% for i = numberOfSamplesForTraining + 1 : M
%     xTesting = [xTesting, xapp(:, p(i))]; % Get testing data sets
%     yTesting = [yTesting, yapp(p(i))]; % Get testing labels
% end
% xTesting = xTesting'; % Transpose
% yTesting = yTesting'; % Transpose
% 
% [numVec, numDim] = size(xTraining);
% avgX = mean(xTraining); % Return the average of training data sets
% stdX = std(xTraining); % Return the standard deviation of training data sets
% disp(avgX);
% disp(stdX);
% for i = 1 : numVec
%     xTraining(i, :) = (xTraining(i, :) - avgX) ./ stdX; % Normal divide
% end
% 
% [numVec, numDim] = size(xTesting);
% for i = 1 : numVec
%     xTesting(i, :) = (xTesting(i, :) - avgX) ./ stdX; % Normal divide
% end
% 
% % Search for the optimal C and gamma, K(x1,x2) = exp{-||x1-x2||^2/gamma}
% % Firstly, search C and gamma in a crude scale
% CScale = [-5, -3, -1, 1, 3, 5, 7, 9, 11, 13, 15];
% gammaScale = [-15, -13, -11, -9, -7, -5, -3, -1, 1, 3];
% C = 2 .^ CScale;
% gamma = 2 .^ gammaScale;
% maxRecognitionRate = 0;
% for i = 1 : length(C)
%     for j = 1 : length(gamma)
%         % 5-fold cross validation
%         cmd = ['-t 2 -c ', num2str(C(i)), ' -g ', num2str(gamma(j)), ' -v 5 -q'];
%         recognitionRate = libsvmtrain(yTraining, xTraining, cmd);
%         if recognitionRate > maxRecognitionRate
%             maxRecognitionRate = recognitionRate;
%             maxCIndex = i;
%             maxGammaIndex = j;
%         end
%     end
% end
% 
% % Then search for optimal C and gamma in a refined scale. 
% n = 10;
% minCScale = 0.5 * (CScale(max(1, maxCIndex - 1)) + CScale(maxCIndex));
% maxCScale = 0.5 * (CScale(min(length(CScale), maxCIndex + 1)) + CScale(maxCIndex));
% newCScale = [minCScale : (maxCScale - minCScale) / n : maxCScale];
% 
% minGammaScale = 0.5 * (gammaScale(max(1, maxGammaIndex - 1)) + gammaScale(maxGammaIndex));
% maxGammaScale = 0.5 * (gammaScale(min(length(gammaScale), maxGammaIndex + 1)) + gammaScale(maxGammaIndex));
% newGammaScale = [minGammaScale : (maxGammaScale - minGammaScale) / n : maxGammaScale];
% newC = 2 .^ newCScale;
% newGamma = 2 .^ newGammaScale;
% maxRecognitionRate = 0;
% for i = 1 : length(newC)
%     for j = 1 : length(newGamma)
%         % 5-fold cross validation
%         cmd = ['-t 2 -c ', num2str(newC(i)), ' -g ', num2str(newGamma(j)), ' -v 5 -q'];
%         recognitionRate = libsvmtrain(yTraining, xTraining, cmd);
%         if recognitionRate > maxRecognitionRate % Iterate and find the maximum rate
%             maxRecognitionRate = recognitionRate
%             maxC = newC(i);
%             maxGamma = newGamma(j);
%         end
%     end
% end
% 
% % Train the SVM model by the optimal C and gamma.
% cmd = ['-t 2 -c ', num2str(maxC), ' -g ', num2str(maxGamma)];
% model = libsvmtrain(yTraining, xTraining, cmd);
% save model.mat model;
% 
% % Test the model on the remaining testing data and obtain the recognition rate.
% load model.mat;
% [yPred, accuracy, decisionValues] = libsvmpredict(yTesting, xTesting, model); 
% save yPred.mat yPred;
% save decisionValues.mat decisionValues;
% save xTraining.mat xTraining;
% save yTesting.mat yTesting;

% draw ROC
[totalScores, index]  = sort(decisionValues);
disp(decisionValues(1:10));
disp(totalScores(1:10))
labels = yTesting;
for i = 1 : length(labels)
    labels(i) = yTesting(index(i));
end

truePositive = zeros(1, length(totalScores) + 1);
trueNegative = zeros(1, length(totalScores) + 1);
falsePositive = zeros(1, length(totalScores) + 1);
falseNegative = zeros(1, length(totalScores) + 1);

for i = 1 : length(totalScores)
    if labels(i) == 1
        truePositive(1) = truePositive(1) + 1;
    else
        falsePositive(1) = falsePositive(1) + 1;
    end
end

for i = 1 : length(totalScores)
   if labels(i) == 1
       truePositive(i + 1) = truePositive(i) - 1;
       falsePositive(i + 1) = falsePositive(i);
   else
       falsePositive(i + 1) = falsePositive(i) - 1;
       truePositive(i + 1) = truePositive(i);
   end
end
truePositive = truePositive / truePositive(1);
falsePositive = falsePositive / falsePositive(1);
plot(falsePositive, truePositive);

File: 例程/兵王问题SVM程序/testSVMChessLibSVM.py
#!/usr/bin/env python3
from libsvm.svmutil import *
from operator import itemgetter
import numpy as np
import random
import matplotlib.pyplot as plt

def readData(filename): # 读取数据
    xApp = []
    yApp = []
    with open(filename, 'rb') as f:
        data = f.readlines()
    for l in data:
        t = l.split(b',')
        # 二分类问题
        if t[-1].startswith(b'draw'):
            y = 0 # 平局为0
        else:
            y = 1 # 胜出为1
        del t[-1]
        # 把字母转化为数字
        xs = [int(c) if ord(c) < 0x3a and ord(c) > 0x2f else ord(c) - ord('a') for c in t]
        xApp.append(xs)
        yApp.append(y)
    return yApp, xApp

def dealWithData(yApp, xApp, trainingDataLength): # 处理数据
    xTraining = []
    yTraining = []
    xTesting = []
    yTesting = []
    idxs = list(range(len(xApp)))
    random.shuffle(idxs) # 打乱数据
    for i in range(trainingDataLength):
        xTraining.append(xApp[idxs[i]])
        yTraining.append(yApp[idxs[i]])
    for i in range(trainingDataLength, len(xApp)):
        xTesting.append(xApp[idxs[i]])
        yTesting.append(yApp[idxs[i]])
    avgX = np.mean(np.mat(xTraining), axis=0).tolist()[0] # 计算训练数据集各个维度的算术平均值
    stdX = np.std(np.mat(xTraining), axis=0).tolist()[0] # 计算训练数据集各个维度的标准方差
    print('[*] avgX = ' + str(avgX))
    print('[*] stdX = ' + str(stdX))
    # 样本归一化
    for data in xTraining:
        for i in range(len(data)):
            data[i] = (data[i] - avgX[i]) / stdX[i]
    for data in xTesting:
        for i in range(len(data)):
            data[i] = (data[i] - avgX[i]) / stdX[i]
    return yTraining, xTraining, yTesting, xTesting

def trainingModel(label, data, modelFilename): # 训练模型
    CScale = [-5, -3, -1, 1, 3, 5, 7, 9, 11, 13, 15]
    gammaScale = [-15, -13, -11, -9, -7, -5, -3, -1, 1, 3]
    maxRecognitionRate = 0
    maxC = 0
    maxGamma = 0
    for C in CScale:
        C_ = pow(2, C)
        for gamma in gammaScale:
            gamma_ = pow(2, gamma)
            cmd = '-t 2 -c ' + str(C_) + ' -g ' + str(gamma_) + ' -v 5 -q'
            recognitionRate = svm_train(label, data, cmd)
            # 比较获取准确率最高的C和gamma
            if recognitionRate > maxRecognitionRate:
                maxRecognitionRate = recognitionRate
                maxC = C
                maxGamma = gamma
    n = 10
    minCScale = 0.5 * (min(-5, maxC) + maxC)
    maxCScale = 0.5 * (max(15, maxC) + maxC)
    newCScale = np.arange(minCScale, maxCScale+1, (maxCScale-minCScale)/n)
    print('[*] newCScale = ' + str(newCScale))
    minGammaScale = 0.5 * (min(-15, maxGamma) + maxGamma)
    maxGammaScale = 0.5 * (max(3, maxGamma) + maxGamma)
    newGammaScale = np.arange(minGammaScale, maxGammaScale+1, (maxGammaScale-minGammaScale)/n)
    print('[*] newGammaScale = ' + str(newGammaScale))
    for C in newCScale:
        C_ = pow(2, C)
        for gamma in newGammaScale:
            gamma_ = pow(2, gamma)
            cmd = '-t 2 -c ' + str(C_) + ' -g ' + str(gamma_) + ' -v 5 -q'
            recognitionRate = svm_train(label, data, cmd)
            # 比较获取准确率最高的C和gamma
            if recognitionRate > maxRecognitionRate:
                maxRecognitionRate = recognitionRate
                maxC = C
                maxGamma = gamma
    # 使用最终确定的C和gamma训练模型
    print('[*] maxC = ' + str(maxC))
    print('[*] maxGamma = ' + str(maxGamma))
    C = pow(2, maxC)
    gamma = pow(2, maxGamma)
    cmd = '-t 2 -c ' + str(C) + ' -g ' + str(gamma)
    model = svm_train(label, data, cmd)
    svm_save_model(modelFilename, model)
    return model

def drawROC(yTesting, decisionValues): # 绘制ROC曲线
    values, labels = [list(x) for x in zip(*sorted(zip(decisionValues, yTesting), key=itemgetter(0)))]
    truePositive = [0 for i in range(len(values) + 1)]
    trueNegative = [0 for i in range(len(values) + 1)]
    falsePositive = [0 for i in range(len(values) + 1)]
    falseNegative = [0 for i in range(len(values) + 1)]
    for i in range(len(values)):
        if labels[i] == 1:
            truePositive[0] += 1
        else:
            falsePositive[0] += 1
    for i in range(len(values)):
        if labels[i] == 1:
            truePositive[i + 1] = truePositive[i] - 1
            falsePositive[i + 1] = falsePositive[i]
        else:
            falsePositive[i + 1] = falsePositive[i] - 1
            truePositive[i + 1] = truePositive[i]
    truePositive = (np.array(truePositive) / truePositive[0]).tolist()
    falsePositive = (np.array(falsePositive) / falsePositive[0]).tolist()
    plt.xlabel('False Positive')
    plt.ylabel('True Positive')
    plt.plot(falsePositive, truePositive, color='blue') # ROC
    plt.plot([1,0], [0,1], color='red') # EER
    plt.legend(['ROC', 'EER'])
    plt.show()

if __name__ == '__main__':
    yApp, xApp = readData('krkopt.data')
    yTraining, xTraining, yTesting, xTesting = dealWithData(yApp, xApp, 5000)
    if input('Train or not? (y/n) ') == 'y':
        model = trainingModel(yTraining, xTraining, 'krkopt.model')
    else:
        model = svm_load_model('krkopt.model')
    yPred, accuracy, decisionValues = svm_predict(yTesting, xTesting, model)
    drawROC(yTesting, decisionValues)

